File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreakHelper(string s, unordered_set<string>& wordSet, int start) {
        if (start == s.size()) {
            return true;
        }

        for (int end = start + 1; end <= s.size(); ++end) {
            if (wordSet.find(s.substr(start, end - start)) != wordSet.end() && 
                wordBreakHelper(s, wordSet, end)) {
                return true;
            }
        }

        return false;
    }

    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        return wordBreakHelper(s, wordSet, 0);
    }
};

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        word_set = set(wordDict)
        memo = {}

        def canBreak(start):
            if start == len(s):
                return True
            if start in memo:
                return memo[start]

            for end in range(start + 1, len(s) + 1):
                if s[start:end] in word_set and canBreak(end):
                    memo[start] = True
                    return True

            memo[start] = False
            return False

        return canBreak(0)
